### PR TITLE
Adapt MKL-DNN operators for both imperative and future JIT environment

### PR DESCRIPTION
**Pull request by CaoZhongZ**
First PR for https://github.com/pytorch/pytorch/issues/23657

Summary:
* Prepare MKL-DNN operators for JIT optimization pass, also promote brand DNNL instead of MKL-DNN
* DNNL operators semantics defined as:
  - All exposed DNNL ops accept both CPUTensor and Opaque Tensor, related standard ATen ops remain unchanged.
  - Computational intensive ops (convolution, etc.) which prefer specific data formats (DNNL’s term) will output Opaque Tensor
  - Memory intensive ops, whose performance doesn’t depend on formats, will output Tensor type the same as input
* Expose full DNNL functionality by adding interfaces with DNNL post ops

Signed-off-by: caozhong <zhong.z.cao@intel.com>

### CODE REVIEW

**Comment by CaoZhongZ** [comment]
@jgong5 @Jianhui-Li

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 1]
```diff
@@ -40,53 +76,116 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
...
+        {output_sizes.cbegin(), output_sizes.cend()},
+        y,
+        {stride.begin(), stride.end()},
+        {dilation.begin(), dilation.end()},
+        {padding.begin(), padding.end()},
+        {padding.begin(), padding.end()},
+        groups,
+        attr_,
+        ideep::algorithm::convolution_direct,
+        ideep::prop_kind::forward_inference);
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp]: It is either "forward" or "forward_inference" depending on who calls it, e.g. conv_relu fusion only applicable in inference while conv better in "forward" training mode aligning current code behavior. Suggest to use a flag here.

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 2]
```diff
@@ -40,53 +76,116 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
...
   std::vector<int64_t> input_size{x_dims.cbegin(), x_dims.cend()};
-  std::vector<int64_t> output_sizes =
-      conv_output_size(input_size, kernel_size, padding, stride, dilation);
+  std::vector<int64_t> output_sizes = conv_output_size(
+      input_size, kernel_size, padding, stride, dilation);
 
-  ideep::tensor y;
   if (b.has_value()) {
-    ideep::convolution_forward::compute<AllocForMKLDNN>(
+    ideep::convolution_forward::compute(
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp]: Missing "alloc" here?

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 3]
```diff
@@ -97,56 +196,106 @@ ideep::tensor _mkldnn_conv2d(
...
-      y,
-      {stride.begin(), stride.end()},
-      {dilation.begin(), dilation.end()},
-      {padding.begin(), padding.end()},
-      {padding.begin(), padding.end()},
-      groups,
-      ideep::descriptor_group::attr_t{},
-      ideep::algorithm::convolution_direct,
-      ideep::prop_kind::forward);
+    ideep::convolution_forward::compute(
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp]: Ditto.

**Pull request closed without merging**